### PR TITLE
LibWeb: Parse css math constants

### PR DESCRIPTION
**Pull request by stelar7**
https://drafts.csswg.org/css-values-4/#calc-constants

### CODE REVIEW

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp [hunk 1]
```diff
@@ -7873,6 +7873,34 @@ ErrorOr<OwnPtr<CalculationNode>> Parser::parse_a_calculation(Vector<ComponentVal
...
+                TRY(values.try_append({ TRY(ConstantCalculationNode::create(CalculationNode::ConstantType::Infinity)) }));
+                continue;
+            }
+
+            if (value.token().ident().equals_ignoring_ascii_case("-infinity"sv)) {
+                TRY(values.try_append({ TRY(ConstantCalculationNode::create(CalculationNode::ConstantType::MinusInfinity)) }));
+                continue;
+            }
+
+            // Note: NaN is case sensitive.
```
Comment (konradekk) [inline Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:7897]: (probably should be “case-sensitive”)

File: Userland/Libraries/LibWeb/CSS/StyleValues/CalculatedStyleValue.cpp [hunk 2]
```diff
@@ -390,6 +390,71 @@ ErrorOr<void> InvertCalculationNode::dump(StringBuilder& builder, int indent) co
...
+{
+    switch (m_constant) {
+    case CalculationNode::ConstantType::E:
+        return { Number(Number::Type::Number, M_E) };
+    case CalculationNode::ConstantType::PI:
+        return { Number(Number::Type::Number, M_PI) };
+    case CalculationNode::ConstantType::Infinity:
+        return { Number(Number::Type::Number, NumericLimits<float>::max()) };
+    case CalculationNode::ConstantType::MinusInfinity:
+        return { Number(Number::Type::Number, NumericLimits<float>::lowest()) };
```
Comment (konradekk) [inline Userland/Libraries/LibWeb/CSS/StyleValues/CalculatedStyleValue.cpp:439]: technically `max` and `lowest` do not denote the representation of infinities… or do they? 🤔

(should those be `INFINITY` and `-INFINITY` resp.?)

**Review by konradekk — COMMENTED** [review]
a question (and a nit)

File: Userland/Libraries/LibWeb/CSS/StyleValues/CalculatedStyleValue.cpp [hunk 3]
```diff
@@ -639,6 +639,70 @@ ErrorOr<void> ClampCalculationNode::dump(StringBuilder& builder, int indent) con
...
+    case CalculationNode::ConstantType::E:
+        return "e"_short_string;
+    case CalculationNode::ConstantType::PI:
+        return "pi"_short_string;
+    case CalculationNode::ConstantType::Infinity:
+        return String::from_utf8("infinity"sv);
+    case CalculationNode::ConstantType::MinusInfinity:
+        return String::from_utf8("-infinity"sv);
+    case CalculationNode::ConstantType::NaN:
+        return String::from_utf8("NaN"sv);
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/StyleValues/CalculatedStyleValue.cpp:667]: Nit: Might as well use `"foo"_string` for these.

File: Userland/Libraries/LibWeb/CSS/StyleValues/CalculatedStyleValue.cpp (hunk 2 above)
Comment (AtkinsSJ) [reply Userland/Libraries/LibWeb/CSS/StyleValues/CalculatedStyleValue.cpp:687]: This is tricky... `infinity - 7` should equal `infinity`, not `FLOAT_MAX - 7`. And similarly, `infinity - infinity` should be `NaN`. So using the actual infinity constants feels correct. But also, from what I can tell we *should* clamp the end result of a calculation tree to the range. (And Chrome seems to agree.)

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/CSS/StyleValues/CalculatedStyleValue.cpp (hunk 2 above)
Comment (stelar7) [reply Userland/Libraries/LibWeb/CSS/StyleValues/CalculatedStyleValue.cpp:687]: It feels like we need to track `NaN` and `Infinity` across every node do resolve it correctly, but with this PR i only wanted to parse the constants, we can improve on handling them correctly later

Comment (AtkinsSJ) [reply Userland/Libraries/LibWeb/CSS/StyleValues/CalculatedStyleValue.cpp:687]: Sure. If you can add a FIXME there though that would be good!

**Review by AtkinsSJ — APPROVED** [review]
Great! :+1:

**Pull request merged**